source/opt/folding_rules.cpp

```diff
@@ -1638,8 +1638,12 @@ FloatConstantKind getFloatConstantKind(const analysis::Constant* constant) {
 
     return kind;
   } else if (const analysis::FloatConstant* fc = constant->AsFloatConstant()) {
-    double value = (fc->type()->AsFloat()->width() == 64) ? fc->GetDoubleValue()
-                                                          : fc->GetFloatValue();
+    if (fc->IsZero()) return FloatConstantKind::Zero;
```

Comment:
Would it make sense to add something like GetDoubleValue() to FloatConstant class? As written this code now doesn't support 1.0 for half floats but does support 0.0 which seems somewhat odd.

Reply:
What are suggesting?  We have `GetDoubleValue()` in the FloatContant class.  It is called below.  Or are you suggesting that `GetDoubleValue()` should return the value as a double regardless of the width of the constant?

Reply:
The latter; maybe this requires a new function name.

Reply:
(it might be worth doing this separately from this PR, to get assertion on half float folding fixed first)

Reply:
LLVM APFloat has:
  bool isZero() const;   ... and other special predicates.
  double convertToDouble() conts;
  bool isExactlyValue(double V) const;